ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobPutHandler.java

```diff
@@ -488,6 +490,34 @@ private void verifyChunkAccountAndContainer(String chunkBlobId, BlobProperties s
       }
     }
 
+    /**
+     * Verify that the reserved metadata id for the specified chunkId is same as seen for previous chunks.
+     * Also return the chunk's reserved metadata id.
+     * @param metadata {@link Map} of metadata set in the signed ids.
+     * @param reservedMetadataBlobId Reserved metadata id for the chunks. Can be {@code null}.
+     * @param chunkId The chunk id.
+     * @return The reserved metadata id.
+     * @throws RestServiceException in case of any exception.
+     */
+    private String getAndVerifyReservedMetadataBlobId(Map<String, String> metadata, String reservedMetadataBlobId,
+        String chunkId) throws RestServiceException {
+      String chunkReservedMetadataBlobId = RestUtils.getHeader(metadata, RestUtils.Headers.RESERVED_METADATA_ID, false);
+      if (reservedMetadataBlobId == null && chunkReservedMetadataBlobId == null) {
+        ReservedMetadataIdMetrics.getReservedMetadataIdMetrics(
+            frontendMetrics.getMetricRegistry()).noReservedMetadataForChunkedUploadCount.inc();
+        if (router.getRouterConfig().routerReservedMetadataEnabled) {
+          throw new RestServiceException(String.format("No reserved metadata id present in chunk %s signed url", chunkId),
+              RestServiceErrorCode.BadRequest);
+        }
+      }
+      if (reservedMetadataBlobId != null && !reservedMetadataBlobId.equals(chunkReservedMetadataBlobId)) {
```

Comment:
When router.getRouterConfig().routerReservedMetadataEnabled is true, the logic is right.
But if it's disabled, some case we throw exception but some case we don't when there is mismatch.
1. First chunk doesn't have the metadata blob id, second chunk has the metadata id. --> No exception.
2. first chunk has the meta blob id, second doesn't have --> throw exception.

The routerReservedMetadataEnabled flag doesn't fully control all the "metadata blob" related logic. It's not clear to me our intention when routerReservedMetadataEnabled is disabled.

Reply:
If the first chunk didn't have a metadata blob id, then we would throw an exception on line 509. Once the first chunk is determined to have a metadata blob id, then line 518 will return that metadata blob id.